### ISSUE TITLE
error using CSSTransitionGroup in 0.5.x

### ISSUE THREAD

**Issue by yogthos** [reagent-project/reagent#88]
This works fine with `0.4.3` release:

```clojure
(def ctg (aget js/React "addons" "CSSTransitionGroup"))

(defn home []  
  [ctg {:transitionName "slider"} [:div "foo"] [:div "bar"]])
```

However, with `0.5.0-alpha`, I get the following error:

    Invariant Violation: onlyChild must be passed a children with exactly one child.

**Comment by city41** [comment]
I am looking into this right now. The cause of the issue is the children of the CSSTransitionGroup end up being a JS object that used to be a cljs vector:


![screen shot 2015-01-28 at 3 07 13 pm](https://cloud.githubusercontent.com/assets/141159/5948092/64f19bc2-a6ff-11e4-8991-00f4df5debba.png)




So I'm picking through and trying to find where that happens. It feels like 0.5.0 is sending something into React just slightly off (thought a method expects x parameters when it really expects y? something like that). So I'm hopeful I can get to the bottom soon.

**Comment by city41** [comment]
I need to stop looking into this for now, but should be able to pick it up again later. 

My test app is this:

```clojure
(def ctg (aget js/React "addons" "CSSTransitionGroup"))
(reagent/render-component [ctg {:transitionName "slider"} [:div]] (.getElementById js/document "app"))
```

using Reagent 0.5.0-alpha (not master), and React 0.12.2.

That is all that is needed to repro the bug. In 0.4.3, this works fine and the end result is a span (the ctg) which contains a div. In 0.5.0-alpha, the error yogthos saw happens.

In 0.5.0-alpha, a call to `do-render` ends up sending this into `ReactElement.createElement`

![screen shot 2015-01-28 at 4 55 21 pm](https://cloud.githubusercontent.com/assets/141159/5949791/de674fba-a70e-11e4-8895-31c8b60970cd.png)

Notice the second and third params are a cljs map (`{:transitionName "slider"}`) and a vector (`[:div]`). `createElement` will take the vector and assign it to `props.children`, and that's what sets up things going bad when downstream `this.state.children` is that vector applied to a JS object.

In 0.4.3, these calls to ReactElement.createElement never directly include cljs data structures. Instead it is always a function (param 1) and a JS object (param 2) with `level` and `argv` as its properties, where `argv` is a vector, and `level` increases as you work down through the component.

In other words, in 0.4.3, ReactElement.createElement's arguments is:

```
[
  function(props) {...},
  { argv: cljs.core.PersistentVector, level: 1 }
]
```

Maybe @holmsand has some ideas?

**Comment by holmsand** [comment]
@city41 Your analysis is spot on. Reagent used to detect that something was a React class, and automagically convert arguments to the form expected by React.

That's no longer possible to do reliably, as far as I can tell (since React.isValidClass was deprecated). But I've added a warning if something looks like it was created by React.createClass (i.e if it has a `type` member). If anyone knows a better way, I'd really like to know...

Instead you will have to call the low-level `createElement` directly.

In the future you can use the brand new `adapt-react-class` that I've just pushed to master. Then this

```
(def ctg (aget js/React "addons" "CSSTransitionGroup"))
```

would become

```
(def ctg (reagent/adapt-react-class (aget js/React "addons" "CSSTransitionGroup")))
```

and then you should be able use `ctg` as before (I hope, fingers crossed, etc).

**Comment by city41** [comment]
This is now working great with `0.5.0-alpha2`. Wrapping the ctg in `adapt-react-class` works as advertised.

I expanded my test app to include actual animations to confirm ctg fully works.

**Comment by jaen** [comment]
FWIW I just changed the old approach of juggling `create-element` and `as-element` to using `adapt-react-class` and everything seems to be working for me as well with `0.5.0-alpha2`.

**Comment by mike-thompson-day8** [comment]
Given v0.5.0 delivered this feature, can this issue can now be closed?
